File: javascript/0707-design-linked-list.js
class Node {
    constructor(val) {
        this.val = val;
        this.prev = null;
        this.next = null;
    }
}

class MyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    get(index) {
        let curr = this.head;
        while (curr && index > 0) {
            curr = curr.next;
            index--;
        }
        return curr ? curr.val : -1;
    }

    addAtHead(val) {
        if (this.head === null) {
            this.head = new Node(val);
            this.tail = this.head;
        } else {
            this.head.prev = new Node(val);
            this.head.prev.next = this.head;
            this.head = this.head.prev;
        }
    }

    addAtTail(val) {
        if (this.head === null) {
            this.head = new Node(val);
            this.tail = this.head;
        } else {
            this.tail.next = new Node(val);
            this.tail.next.prev = this.tail;
            this.tail = this.tail.next;
        }
    }

    addAtIndex(index, val) {
        if (index === 0) this.addAtHead(val);
        else {
            let curr = this.head;
            while (curr && index > 0) {
                curr = curr.next;
                index--;
            }
            if (index === 0) {
                if (!curr) this.addAtTail(val);
                else {
                    const prev = curr.prev;
                    prev.next = new Node(val);
                    prev.next.prev = prev;
                    prev.next.next = curr;
                    curr.prev = prev.next;
                }
            }
        }
    }

    deleteAtIndex(index) {
        if (!this.head) return;
        if (index === 0) {
            this.head = this.head.next;
            if (this.head) this.head.prev = null;
        } else {
            let curr = this.head;
            while (curr && index > 0) {
                curr = curr.next;
                index--;
            }
            if (curr && index === 0) {
                if (!curr.next) {
                    curr.prev.next = null;
                    this.tail = curr.prev;
                } else {
                    curr.prev.next = curr.next;
                    curr.next.prev = curr.prev;
                }
            }
        }
    }
}

File: go/0707-design-linked-list.go
type Node struct {
	val  int
	prev *Node
	next *Node
}

func NewNode(val int, prev, next *Node) *Node {
	return &Node{
		val:  val,
		prev: prev,
		next: next,
	}
}

type MyLinkedList struct {
	head *Node
	tail *Node
}

func Constructor() MyLinkedList {
	head := NewNode(-1, nil, nil)
	tail := NewNode(-1, head, nil)
	head.next = tail

	return MyLinkedList{
		head: head,
		tail: tail,
	}
}

func (this *MyLinkedList) Get(index int) int {
	cur := this.head.next
	for cur != nil && index > 0 {
		cur = cur.next
		index--
	}

	if cur != nil && cur != this.tail && index == 0 {
		return cur.val
	}

	return -1
}

func (this *MyLinkedList) AddAtHead(val int) {
	node := NewNode(val, this.head, this.head.next)
	this.head.next.prev = node
	this.head.next = node
}

func (this *MyLinkedList) AddAtTail(val int) {
	node := NewNode(val, this.tail.prev, this.tail)
	this.tail.prev.next = node
	this.tail.prev = node
}

func (this *MyLinkedList) AddAtIndex(index int, val int) {
	cur := this.head.next
	for cur.next != nil && index > 0 {
		cur = cur.next
		index--
	}

	if index == 0 {
		node := NewNode(val, cur.prev, cur)
		cur.prev.next = node
		cur.prev = node
	}
	return
}

func (this *MyLinkedList) DeleteAtIndex(index int) {
	cur := this.head.next

	for cur.next != nil && index > 0 {
		cur = cur.next
		index--
	}

	if index == 0 && cur != nil && cur != this.tail {
		prev, next := cur.prev, cur.next
		prev.next, next.prev = next, prev
	}
	return
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Get(index);
 * obj.AddAtHead(val);
 * obj.AddAtTail(val);
 * obj.AddAtIndex(index,val);
 * obj.DeleteAtIndex(index);
 */
